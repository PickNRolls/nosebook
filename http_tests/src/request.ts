import supertest, { Test } from 'supertest';
import TestAgent from 'supertest/lib/agent';

const HOST = 'http://backend:8080';
const SESSION_HEADER_KEY = 'X-Auth-Session-Id';

class TestAgentWrapper {
  private host: string;
  private agent: TestAgent;

  public constructor(host: string) {
    this.host = host;
    this.agent = supertest(host);
  }

  public unwrap(): TestAgent {
    return this.agent;
  }

  public extend(prefixUrl: string): TestAgentWrapper {
    return new TestAgentWrapper(this.host + prefixUrl);
  }

  private hook<M extends 'get' | 'post'>(method: M, url: string): Test {
    const test = this.agent[method](url);

    test.set(SESSION_HEADER_KEY, 'bb23af03-be50-4bce-b729-b259b2e02e54');

    return test;
  }

  public get(url: string): Test {
    return this.hook('get', url);
  }

  public post(url: string): Test {
    return this.hook('post', url);
  }
};

export default new TestAgentWrapper(HOST);


