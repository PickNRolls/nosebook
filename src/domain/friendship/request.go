package domainfriendship

import (
	"nosebook/src/errors"
	"time"

	"github.com/google/uuid"
)

type FriendRequest struct {
	RequesterId uuid.UUID `json:"requesterId" db:"requester_id"`
	ResponderId uuid.UUID `json:"responderId" db:"responder_id"`
	Message     string    `json:"message" db:"message"`
	Accepted    bool      `json:"accepted" db:"accepted"`
	Viewed      bool      `json:"viewed" db:"viewed"`
	CreatedAt   time.Time `json:"createdAt" db:"created_at"`

	events []Event
}

func New(
	requesterId uuid.UUID,
	responderId uuid.UUID,
	message string,
	accepted bool,
	viewed bool,
	createdAt time.Time,
	raiseCreatedEvent bool,
) *FriendRequest {
	output := &FriendRequest{
		RequesterId: requesterId,
		ResponderId: responderId,
		Message:     message,
		Accepted:    accepted,
		Viewed:      viewed,
		CreatedAt:   createdAt,
		events:      make([]Event, 0),
	}

	if raiseCreatedEvent {
		output.raiseEvent(NewCreatedEvent())
	}

	return output
}

func (this *FriendRequest) raiseEvent(event Event) {
	this.events = append(this.events, event)
}

func (this *FriendRequest) Events() []Event {
	return this.events
}

func (this *FriendRequest) ViewBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отметить заявку просмотренной")
	}

	this.Viewed = true
	this.raiseEvent(NewViewedEvent())

	return nil
}

func (this *FriendRequest) AcceptBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может принять заявку")
	}

	this.Accepted = true
	this.Viewed = true
	this.raiseEvent(NewAcceptedEvent())

	return nil
}

func (this *FriendRequest) DenyBy(userId uuid.UUID) *errors.Error {
	if userId != this.ResponderId {
		return errors.New("Friendship Error", "Только получатель может отклонить заявку")
	}

	this.Accepted = false
	this.Viewed = true
	this.raiseEvent(NewDeniedEvent())

	return nil
}

func (this *FriendRequest) RemoveBy(userId uuid.UUID) *errors.Error {
	if !this.Accepted {
		return errors.New("Friendship Error", "Нельзя удалить непринятую заявку")
	}

	this.Accepted = false
	previousRequesterId := this.RequesterId
	previousResponderId := this.ResponderId

	if this.RequesterId == userId {
		this.RequesterId, this.ResponderId = this.ResponderId, this.RequesterId
	}

	this.raiseEvent(NewRemovedEvent(previousRequesterId, previousResponderId))

	return nil
}
