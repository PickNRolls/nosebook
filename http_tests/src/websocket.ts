import * as ws from 'ws';
import { HOST, SESSION_HEADER_KEY, TEST_TESTER_SESSION } from './request';

export class WebSocket {
  private original: ws.WebSocket;

  public constructor(session?: string) {
    this.original = new ws.WebSocket(`${HOST}/ws`, {
      headers: {
        [SESSION_HEADER_KEY]: session ?? TEST_TESTER_SESSION,
      }
    });
  }

  public unwrap() {
    return this.original;
  }
}

