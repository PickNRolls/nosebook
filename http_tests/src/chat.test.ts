import request, { ASSER_ID, ASSER_SESSION } from "./request";
import { WebSocket } from "./websocket";

describe('posts', () => {
  let chats = request.extend({ prefixUrl: '/chats' });


  describe('GET /', () => {
    test('result list of 2 chats', async () => {
      let response = await chats
        .get('')
        .expect(200)

      expect(response.body).toMatchSnapshot();
    });

    test('filter limit = 1', async () => {
      let response = await chats
        .get('')
        .query({
          limit: 1,
        })
        .expect(200)

      expect(response.body).toMatchSnapshot();

      response = await chats
        .get('')
        .query({
          limit: 1,
          next: response.body.data.next,
        })
        .expect(200);

      expect(response.body).toMatchSnapshot();
    });
  });

  const conv = request.extend({ prefixUrl: '/conversations' });

  test('POST /send-message', async () => {
    const websocket = new WebSocket(ASSER_SESSION).unwrap();

    const message = new Promise(res => {
      websocket.once('message', (data) => {
        res(JSON.parse(data.toString()));
      });
    });

    let response = await conv
      .post('/send-message')
      .send({
        recipientId: ASSER_ID,
        text: 'Hello asser'
      })
      .expect(200);

    expect(response.body).toMatchSnapshot();
    expect(await message).toMatchSnapshot();

    websocket.terminate();
  });
});

