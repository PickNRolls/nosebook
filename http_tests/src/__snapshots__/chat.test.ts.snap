// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts GET / filter limit = 1 1`] = `
{
  "data": {
    "data": [
      {
        "createdAt": "2024-08-10T10:06:02Z",
        "id": "ad9ae3dd-9a07-4d68-a66e-db721928e8dd",
        "interlocutor": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "lastMessage": {
          "author": {
            "firstName": "Test",
            "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
            "lastName": "Tester",
            "lastOnlineAt": "2024-08-10T10:10:10Z",
            "nick": "test_tester",
            "online": true,
          },
          "chatId": "ad9ae3dd-9a07-4d68-a66e-db721928e8dd",
          "createdAt": "2024-08-10T10:06:19Z",
          "id": "e5bba43d-2888-4bb9-b5be-9c60face9331",
          "text": "Hello asser sample message 2",
        },
      },
    ],
    "next": "ad9ae3dd-9a07-4d68-a66e-db721928e8dd/2024-08-10T10:06:19Z",
    "totalCount": 2,
  },
  "ok": true,
}
`;

exports[`posts GET / filter limit = 1 2`] = `
{
  "data": {
    "data": [
      {
        "createdAt": "2024-08-10T10:06:02Z",
        "id": "ad9ae3dd-9a07-4d68-a66e-db721928e8da",
        "interlocutor": {
          "firstName": "Marina",
          "id": "48683858-796c-45ad-a361-9e3d6d003354",
          "lastName": "Graf",
          "lastOnlineAt": "2024-08-10T10:10:02Z",
          "nick": "mmm",
          "online": true,
        },
        "lastMessage": {
          "author": {
            "firstName": "Test",
            "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
            "lastName": "Tester",
            "lastOnlineAt": "2024-08-10T10:10:10Z",
            "nick": "test_tester",
            "online": true,
          },
          "chatId": "ad9ae3dd-9a07-4d68-a66e-db721928e8da",
          "createdAt": "2024-08-10T10:06:19Z",
          "id": "e5bba43d-2888-4bb9-b5be-9c60face9332",
          "text": "Hello marina sample message",
        },
      },
    ],
    "totalCount": 2,
  },
  "ok": true,
}
`;

exports[`posts GET / result list of 2 chats 1`] = `
{
  "data": {
    "data": [
      {
        "createdAt": "2024-08-10T10:06:02Z",
        "id": "ad9ae3dd-9a07-4d68-a66e-db721928e8dd",
        "interlocutor": {
          "firstName": "Ilya",
          "id": "1ae02f69-ea1a-4308-b825-0e5896e652e4",
          "lastName": "Blinkov",
          "lastOnlineAt": "2024-08-10T10:01:02Z",
          "nick": "drugtester",
          "online": false,
        },
        "lastMessage": {
          "author": {
            "firstName": "Test",
            "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
            "lastName": "Tester",
            "lastOnlineAt": "2024-08-10T10:10:10Z",
            "nick": "test_tester",
            "online": true,
          },
          "chatId": "ad9ae3dd-9a07-4d68-a66e-db721928e8dd",
          "createdAt": "2024-08-10T10:06:19Z",
          "id": "e5bba43d-2888-4bb9-b5be-9c60face9331",
          "text": "Hello asser sample message 2",
        },
      },
      {
        "createdAt": "2024-08-10T10:06:02Z",
        "id": "ad9ae3dd-9a07-4d68-a66e-db721928e8da",
        "interlocutor": {
          "firstName": "Marina",
          "id": "48683858-796c-45ad-a361-9e3d6d003354",
          "lastName": "Graf",
          "lastOnlineAt": "2024-08-10T10:10:02Z",
          "nick": "mmm",
          "online": true,
        },
        "lastMessage": {
          "author": {
            "firstName": "Test",
            "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
            "lastName": "Tester",
            "lastOnlineAt": "2024-08-10T10:10:10Z",
            "nick": "test_tester",
            "online": true,
          },
          "chatId": "ad9ae3dd-9a07-4d68-a66e-db721928e8da",
          "createdAt": "2024-08-10T10:06:19Z",
          "id": "e5bba43d-2888-4bb9-b5be-9c60face9332",
          "text": "Hello marina sample message",
        },
      },
    ],
    "totalCount": 2,
  },
  "ok": true,
}
`;

exports[`posts POST /send-message 1`] = `
{
  "data": true,
  "ok": true,
}
`;

exports[`posts POST /send-message 2`] = `
{
  "payload": {
    "author": {
      "firstName": "Test",
      "id": "ed1a3fd0-4d0b-4961-b4cd-cf212357740d",
      "lastName": "Tester",
      "lastOnlineAt": "2024-08-10T10:10:10Z",
      "nick": "test_tester",
      "online": true,
    },
    "chatId": "test chat id",
    "createdAt": "2024-08-10T10:10:10Z",
    "id": "test id",
    "text": "Hello asser",
  },
  "type": "new_message",
}
`;
